ulib/LowStar.ConstBuffer.fsti

```diff
@@ -0,0 +1,187 @@
+   the weakest (i.e., mutability).
+
+   The main type of this module is `const_buffer t`. It is extracted
+   by KreMLin to  `const t*`, i.e., a
+*)
+
+module P = FStar.Preorder
+module G = FStar.Ghost
+module U32 = FStar.UInt32
+module Seq = FStar.Seq
+
+module HS = FStar.HyperStack
+open FStar.HyperStack.ST
+
+module I = LowStar.ImmutableBuffer
+module B = LowStar.Buffer
+
+(*** A model for const pointers **)
+
+/// We start by defining the finite disjunction of mutable and
+/// immutable buffers.
+///
+/// NB: THIS IS FOR SPECIFICATIONAL PURPOSES ONLY
+/// The concrete type `const_buffer` is defined later
+
+/// `qual`: mutability qualifier
+type qual =
+  | MUTABLE
+  | IMMUTABLE
+
+/// `qbuf_cases q a`: disjunction of mutable and immutable pointers
+let qbuf_cases (q:qual) (a:Type) =
+  match q with
+  | MUTABLE -> B.buffer a
+  | IMMUTABLE -> I.ibuffer a
+
+/// `q_preorder q a`: As we'll see shortly, it is convenient to also
+/// define a disjunction of the preorder indices on a qualified
+/// buffer
+let q_preorder (q:qual) (a:Type) =
+  match q with
+  | MUTABLE -> B.trivial_preorder a
+  | IMMUTABLE -> I.immutable_preorder a
+
+/// `qbuf a`: This type is used for specificational purposes only. It
+///  is buffer whose mutability qualifier is existentially bound, via
+///  a dependent pair
+let qbuf a = (q:qual & qbuf_cases q a)
+
+/// `qbuf_qual c`: projecting the mutability qualifier of a `qbuf`
+let qbuf_qual (c:qbuf 'a) : qual = dfst c
+
+/// `qbuf_pre c`: case-dependent preorders for a qbuf
+let qbuf_pre (c:qbuf 'a)  = q_preorder (qbuf_qual c) 'a
+
+/// `qbuf_mbuf c`: turning a qbuf into a regular monotonic buffer
+let qbuf_mbuf (c:qbuf 'a) : B.mbuffer 'a (qbuf_pre c) (qbuf_pre c) = dsnd c
+
+(*** CONCRETE CONST POINTERS **)
+
+/// `const_buffer`:
+///  An abstract type of a read-only pointer to an array of `a`
+val const_buffer (a:Type u#0) : Type u#0
+
+
+/// `as_qbuf`: For specificational purposes, a const_buffer can be
+/// seen as an existentially quantified qbuf
+val as_qbuf (c:const_buffer 'a) : GTot (qbuf 'a)
+
+/// `as_mbuf`: A convenience function to turn a const_buffer into a
+/// regular mbuffer, for spec purposes
+let as_mbuf (c:const_buffer 'a) : GTot _ = qbuf_mbuf (as_qbuf c)
+
+/// We now give several convenience functions that lift common
+/// notions on buffers to const_buffer, via the `as_mbuf` coercion
+let live (h:HS.mem) (c:const_buffer 'a)   = B.live h (as_mbuf c)
+let length (c:const_buffer 'a)            = B.length (as_mbuf c)
+let loc_buffer (c:const_buffer 'a)        = B.loc_buffer (as_mbuf c)
+let as_seq (h:HS.mem) (c:const_buffer 'a) = B.as_seq h (as_mbuf c)
+
+(*** CONSTRUCTORS **)
+
+/// `of_buffer`: A constructors for const buffers from mutable and
+/// immutable buffers. It is fully specified in terms of the
+/// `qbuf/mbuf` model
+val of_buffer (b:B.buffer 'a)
+  : Pure (const_buffer 'a)
+    (requires True)
+    (ensures fun c ->
```

Comment:
Small comment, but I guess you could write the postcondition more succinctly as `as_qbuf c == (| MUTABLE, b |)`? And same for the next function.

Reply:
I actually prefer not to do that. 

The reason is that the other specs (e.g., the liveness precondition of `index`) are written (and must be so) in terms of as_qbuf, qbuf_mbuf etc. 

So it's more uniform, especially for the SMT solver, to have all specs to use the same functions, rather than switching between styles.